File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        temp = -1000000
        def doMax(root):
            nonlocal temp
            if root is None:
                return 0
            maxLeft = doMax(root.left)
            maxRight = doMax(root.right)
            temp = max([maxLeft + root.val, maxRight + root.val, maxLeft + maxRight + root.val, root.val, temp])
            return max([maxLeft + root.val, maxRight + root.val, root.val])

        doMax(root)
        return temp

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    int numGood{0};
public:
    void doGood(TreeNode *root, int max) {
        if (!root) return;
        if (root->val >= max) numGood++;
        max = std::max(max, root->val);
        doGood(root->left, max);
        doGood(root->right, max);
    }
    int goodNodes(TreeNode* root) {
        doGood(root, -1000);
        return numGood;
    }
};
